### PR TITLE
Refactor python module installer and add install-python target to CMake

### LINKED ISSUE #1373: [Feature Request] Implement equivalent to 'make install-python' in CMake
**Summary**

After building the LAMMPS shared library via CMake there should be a way to install the python module similar as with the conventional build system.

**Detailed Description**

The conventional build system allows to install the python module with the matching shared library through "make install-python", either into the system or user python site-packages folder. There doesn't seem to be something equivalent for the CMake build. 'make install' just installs executable, library, potentials, environment, manpage, cmake and pkgconfig files.

### PR DESCRIPTION
**Pull request by akohlmey**
**Summary**

This pull request provides some refactoring of the LAMMPS python module installation script to improve platform neutrality and include it cleanly into both, conventional and CMake based builds

**Related Issues**

closes #1373 

**Author(s)**

Axel Kohlmeyer (Temple U)

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**

The command line flags for manually running the Python module install script have changed. However suitable help messages and consistency checks on arguments have been added.

**Implementation Notes**

The CMake old configuration made (part of) the python module installation dependent on enabling the PYTHON package. This makes this independent and CMake builds should behave similar to conventional make now.

**Post Submission Checklist**

_Please check the fields below as they are completed **after** the pull request has been submitted. Delete lines that don't apply_

- [x] The feature or features in this pull request is complete
- [x] Licensing information is complete
- [x] Corresponding author information is complete
- [x] The source code follows the LAMMPS formatting guidelines
- [x] Suitable new documentation files and/or updates to the existing docs are included
- [x] The added/updated documentation is integrated and tested with the documentation build system
- [x] The feature has been verified to work with the conventional build system
- [x] The feature has been verified to work with the CMake based build system
- [ ] A package specific README file has been included or updated
- [ ] One or more example input decks are included

**Further Information, Files, and Links**

_Put any additional information here, attach relevant text or image files, and URLs to external sites (e.g. DOIs or webpages)_

### CODE REVIEW

**Comment by akohlmey** [comment]
> 2 questions:
> 
> * Does this support `DESTDIR`, e.g. `make install-python DESTDIR=/some/packaging/path` - rpm is using that.

if you want to install into a custom location, you need to call python/install.py directly and set the -d flag

> * How to pick a non-standard install location, e.g. on Fedora python packages for OpenMPI need to go into `$MPI_PYTHON3_SITEARCH`, which is something like `/usr/lib64/python3.X/openmpi/`.
> 
> In general, I don't like the idea of one buildsystem (CMake) calling another buildsystem (`python-install.py`).

what has been there before was not working at all and incorrectly tied the LAMMPS python module to the PYTHON package. What this PR adds is a consistent behavior and default settings that work in both build systems for people that compile LAMMPS from source. This is the primary audience and not people packaging LAMMPS. Since packaging is scripted anyway, it should be easy to customize installation any which way you like in the packaging scripts.

using distutils is what i thought to be the "proper way" to install python modules. now both build systems do it.

**Comment by junghans** [comment]
Ok, I still don't like the fact that we are breaking with the standard CMake workflow, i.e. `cmake ; make ; make install`.

File: cmake/CMakeLists.txt [hunk 1]
```diff
@@ -1483,6 +1474,26 @@ install(
...
+# Install LAMMPS module into site-packages folder
+# Only available, if a shared library is built
+###############################################################################
+if(BUILD_LIB AND BUILD_SHARED_LIBS)
+  find_package(PythonInterp)
+  add_custom_target(
+    install-python
+    ${PYTHON_EXECUTABLE} install.py -v ${LAMMPS_SOURCE_DIR}/version.h
+    -m ${CMAKE_CURRENT_SOURCE_DIR}/../python/lammps.py
+    -l ${CMAKE_BINARY_DIR}/liblammps${CMAKE_SHARED_LIBRARY_SUFFIX}
```
Comment (junghans) [inline cmake/CMakeLists.txt]: Could we add `-d ${PYTHON_INSTDIR}` here?

File: cmake/CMakeLists.txt [hunk 2]
```diff
@@ -1483,6 +1474,26 @@ install(
   DESTINATION ${CMAKE_INSTALL_SYSCONFDIR}/profile.d
 )
 
+###############################################################################
+# Install LAMMPS module into site-packages folder
+# Only available, if a shared library is built
+###############################################################################
+if(BUILD_LIB AND BUILD_SHARED_LIBS)
```
Comment (junghans) [inline cmake/CMakeLists.txt]: Should that be behind `if(PKG_PYTHON)`? Also I was always wondering if `PKG_PYTHON` even made sense with `BUILD_LIB=OFF` or `BUILD_SHARED_LIBS=OFF`.

**Comment by junghans** [comment]
Please correct me if I am wrong, but the only thing `install.py` does, is to install `liblammps.so` in the site-package directory a second time and to create an egg file, correct?

Even though I have never seen there egg files uses for anything, we could easily create them from a template (like our pkg-config file) as the only part, which actually changes is the version string.

**Comment by akohlmey** [comment]
@sjplimp do we need to keep the legacy installation scheme for the python module, that just copies the LAMMPS python module and the shared library to a given folder and then requires the user to explicitly set environment variables? I am thinking that if a user will set the environment variables, then they can point to the source (like you seem to prefer doing).

I am trying to make this consistent across platforms (Windows, MacOSX, and Linux all have different distribution trees and - sometimes - shared library extensions) and have this consistent across cmake and make build (so that something meaningful happens with "make install-python" in either case and whether it is called with superuser permissions or not) and calling the install script manually.

File: cmake/CMakeLists.txt (hunk 2 above)
Comment (akohlmey) [reply cmake/CMakeLists.txt]: `PKG_PYTHON` and the LAMMPS python module are two different things. The python module provides a LAMMPS module to python and uses ctypes to interface functions from the library interface imported from a shared library. The PYTHON package embeds the python interpreter into LAMMPS. Now it **is** possible to use both at the same time, but that is rather the exception than the rule.
Yes, `PKG_PYTHON` makes a lot of sense without building libraries. I do it all the time.

TL;DR: `PKG_PYTHON` allows to run python code inside of LAMMPS, `lammps.py` allows to run LAMMPS inside of python.

**Comment by akohlmey** [comment]
> Ok, I still don't like the fact that we are breaking with the standard CMake workflow, i.e. `cmake ; make ; make install`.

this builds/installs LAMMPS, the executable.

LAMMPS, the python module is a different thing. We have plenty of people that only want the latter.

**Comment by akohlmey** [comment]
@junghans i've studied the distutils/setuptools docs, and there is a `prefix` option, which i have tried to expose to the `install.py` script. Would this be compatible with the installation tree requirements of Fedora?

What I want to bring to the CMake setup is, that `make install-python` will produce a working python module regardless of having root access or not. It will first try to install into the standard system tree, and if that fails (usually due to insufficient permissions), it will re-run the installation attempt with the `--user` flag added. After that the python module install should just work (well, at least on most Linux systems and MacOSX) without having to mess with environment variables.

File: cmake/CMakeLists.txt (hunk 1 above)
Comment (junghans) [reply cmake/CMakeLists.txt]: With the recent change, should we are `-p` here? CMake's default prefix already doesn't require root.

Comment (akohlmey) [reply cmake/CMakeLists.txt]: I am OK with any change, that keeps the conventional build and cmake builds with default settings consistent. E.g. on MacOSX the user site-packages folder for python is "$HOME/Library/Python/#.#/lib/python/site-packages". Does `${PYTHON_INSTDIR}` by default point there? While on Linux it would have to point to "$HOME/.local/lib/python#.#/site-packages/" by default.

File: cmake/CMakeLists.txt [hunk 3]
```diff
@@ -1483,6 +1474,26 @@ install(
...
+  find_package(PythonInterp)
```
Comment (junghans) [inline cmake/CMakeLists.txt]: Missing `REQUIRED` or put the new command inside `if(PHYTHON_FOUND)`.

File: cmake/CMakeLists.txt (hunk 1 above)
Comment (junghans) [reply cmake/CMakeLists.txt]: It runs `import distutils.sysconfig as cg; print(cg.get_python_lib(1,0,prefix='${CMAKE_INSTALL_PREFIX}))` to determine the library dir.

**Comment by akohlmey** [comment]
@junghans it look like there are too many conflicts in how `make install` and `make install-python` is expected to work. however, earlier today i realized, that the common usage scenarios for those two targets and pretty orthogonal, i.e. a user that would want to do `make install-python` would not likely be interested in also doing `make install` and vice versa. So the resolution could be to install the python module with *both* targets, but with the different settings. For `make install` it makes sense to install the module when *either* the PYTHON package is enabled (as you can pass a reference to SELF which dlsym() will resolve properly even without a shared library, if LAMMPS has been linked to re-export all symbols, which is enabled by the linker flags of the python libraries) *or* the shared library is built. When neither is enabled, there is no way to initialize the python module, so `make install` will not copy `lammps.py`.

File: cmake/CMakeLists.txt [hunk 4]
```diff
@@ -1475,23 +1475,47 @@ install(
...
   find_package(PythonInterp)
-  add_custom_target(
-    install-python
-    ${PYTHON_EXECUTABLE} install.py -v ${LAMMPS_SOURCE_DIR}/version.h
-    -m ${CMAKE_CURRENT_SOURCE_DIR}/../python/lammps.py
-    -l ${CMAKE_BINARY_DIR}/liblammps${CMAKE_SHARED_LIBRARY_SUFFIX}
-    WORKING_DIRECTORY  ${CMAKE_CURRENT_SOURCE_DIR}/../python
-    COMMENT "Installing LAMMPS Python module"
-    )
+  if (${PYTHONINTERP_FOUND})
```
Comment (junghans) [inline cmake/CMakeLists.txt]: `${}` not needed.

File: cmake/CMakeLists.txt [hunk 5]
```diff
@@ -1483,6 +1474,50 @@ install(
...
+      install-python
+      ${PYTHON_EXECUTABLE} install.py -v ${LAMMPS_SOURCE_DIR}/version.h
+      -m ${CMAKE_CURRENT_SOURCE_DIR}/../python/lammps.py
+      -l ${CMAKE_BINARY_DIR}/liblammps${CMAKE_SHARED_LIBRARY_SUFFIX}
+      WORKING_DIRECTORY  ${CMAKE_CURRENT_SOURCE_DIR}/../python
+      COMMENT "Installing LAMMPS Python module")
+  else()
+    add_custom_target(
+      install-python
+      echo "Must have Python installed to install the LAMMPS Python module")
```
Comment (junghans) [inline cmake/CMakeLists.txt]: Nitpick: I would use `${CMAKE_COMMAND} -E echo` for better Cross-Plattform support.

File: cmake/CMakeLists.txt [hunk 6]
```diff
@@ -1483,6 +1474,50 @@ install(
...
+  endif()
+else()
+  add_custom_target(
+    install-python
+    echo "Must build LAMMPS as a shared library to use the Python module")
```
Comment (junghans) [inline cmake/CMakeLists.txt]: Same nitpick!

File: cmake/CMakeLists.txt [hunk 7]
```diff
@@ -1483,6 +1474,50 @@ install(
...
+endif()
+
+###############################################################################
+# Add LAMMPS python module to "install" target. This is taylored for building
+# LAMMPS for package managers and with different prefix settings.
+# This requires either a shared library or that the PYTHON package is included.
+###############################################################################
+if((BUILD_LIB AND BUILD_SHARED_LIBS) OR (PKG_PYTHON))
+  find_package(PythonInterp)
+  if (${PYTHONINTERP_FOUND})
```
Comment (junghans) [inline cmake/CMakeLists.txt]: Same for `${}`.

File: cmake/CMakeLists.txt [hunk 8]
```diff
@@ -1483,6 +1474,50 @@ install(
...
+    if(NOT PYTHON_INSTDIR)
```
Comment (junghans) [inline cmake/CMakeLists.txt]: `PYTHON_INSTDIR` is not a real cached variable maybe it is time to make it one and have it default to the value below.

**Comment by rbberger** [comment]
I don’t have time to test this, but does this work with virtualenv?

**Comment by akohlmey** [comment]
On Wed, Mar 27, 2019 at 11:26 AM Richard Berger <notifications@github.com>
wrote:

> I don’t have time to test this, but does this work with virtualenv?
>
CMake needs a tweak if you ran it outside  of virutalenv and then do make
python-install inside, as it hardcodes the path to the python executable. i've
updated the pylammps docs accordingly, which is the only place in the
manual referencing virtualenv

conventional make picks up the different python executable in virtualenv
automatically.

axel.

>
